Reuse TrackMetadata in TrackInfoObject

uklotzde opened the pull request:
Get rid of many redundant member variables in TrackInfoObject by replacing them with TrackMetadata.

uklotzde committed:
Reuse TrackMetadata in TrackInfoObject

uklotzde committed:
Rename boolean getter function

daschuer commented on src/trackinfoobject.cpp:
```diff
@@ -147,28 +161,10 @@ void TrackInfoObject::setMetadata(const Mixxx::TrackMetadata& trackMetadata) {
+void TrackInfoObject::getTrackMetadata(Mixxx::TrackMetadata* pTrackMetadata, bool* pParsedFromFile) const {
+    QMutexLocker lock(&m_qMutex);
+    *pTrackMetadata = m_metadata;
+    *pParsedFromFile = m_bHeaderParsed;
```
Now the pParsedFromFile flag looks like to indicate that the Metadata is 100 % from the file. 

"m_bHeaderParsed" distingushes a file with initial metadata from a the file, from a file that has no file meta data. 
Even if a users has changes all metadata, m_bHeaderParsed is still true.  

How about just keep the name  "pHeaderParsed" to match the library name or something like "pParsedFromFileBefore"?

uklotzde committed:
Consistent naming of headerParsed flag

daschuer commented on src/soundsourceproxy.cpp:
```diff
@@ -273,8 +273,8 @@ SoundSourceProxy::SaveTrackMetadataResult SoundSourceProxy::saveTrackMetadata(
-        bool parsedFromFile = pTrack->getHeaderParsed();
-        pTrack->getMetadata(&trackMetadata);
+        bool parsedFromFile = false;
+        pTrack->getTrackMetadata(&trackMetadata, &parsedFromFile);
```
Not sure if here is a good place for CPU optimization, but at least is it possible. 
This function It is called in sometimes in a a loop like in the browse feature, 

How about making trackMetadata a pointer member of pTrack? This would allow us to get around some extra deep copies. We basically need only one deep copy, when creating the metadata clone.
Even the default constructor can be optimized if we return a copy constructed metadata object from    getTrackMetadata()

A separate bool parsedFromFile = pTrack->getHeaderParsed(); would allow us to return without copy the metadata content,

uklotzde replied:
That would violate the thread-safety guarantees of TrackInfoObject. We need 
to return a copy of the data, not only a pointer to internal data that might 
be modified at any time by another thread.

Copying some primitive values and copy-on-write QString objects is cheap 
compared to locking and unlocking the TIO multiple times.

On 02/28/2016 09:52 PM, Daniel Schürmann wrote:
>
> In src/soundsourceproxy.cpp 
> <https://github.com/mixxxdj/mixxx/pull/900#discussion_r54356078>:
>
> > @@ -273,8 +273,8 @@ SoundSourceProxy::SaveTrackMetadataResult SoundSourceProxy::saveTrackMetadata(
> >      SoundSourceProxy proxy(pTrack);
> >      if (!proxy.m_pSoundSource.isNull()) {
> >          Mixxx::TrackMetadata trackMetadata;
> > -        bool parsedFromFile = pTrack->getHeaderParsed();
> > -        pTrack->getMetadata(&trackMetadata);
> > +        bool parsedFromFile = false;
> > +        pTrack->getTrackMetadata(&trackMetadata, &parsedFromFile);
>
> Not sure if here is a good place for CPU optimization, but at least is it 
> possible.
> This function It is called in sometimes in a a loop like in the browse 
> feature,
>
> How about making trackMetadata a pointer member of pTrack? This would 
> allow us to get around some extra deep copies. We basically need only one 
> deep copy, when creating the metadata clone.
> Even the default constructor can be optimized if we return a copy 
> constructed metadata object from getTrackMetadata()
>
> A separate bool parsedFromFile = pTrack->getHeaderParsed(); would allow us 
> to return without copy the metadata content,
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/mixxxdj/mixxx/pull/900/files#r54356078>.
>

daschuer replied:
Ah, thank you for the reminder.

daschuer commented on src/trackinfoobject.cpp:
```diff
@@ -110,26 +101,49 @@ void TrackInfoObject::setDeleteOnReferenceExpiration(bool deleteOnReferenceExpir
+                (m_metadata.getReplayGain() != trackMetadata.getReplayGain());
+        bool modifiedCoverArt = false;
+        if (m_metadata != trackMetadata) {
+            const Mixxx::ReplayGain replayGainBackup(m_metadata.getReplayGain());
```
Why we need a special treatment for replay gain here? Don't we trust the file replay gain?

uklotzde replied:
Good point. Looks like the reason for this special case handling got lost on 
the way. I have to think about it again ;)

On 02/28/2016 10:15 PM, Daniel Schürmann wrote:
>
> In src/trackinfoobject.cpp 
> <https://github.com/mixxxdj/mixxx/pull/900#discussion_r54356482>:
>
> > -    setBitrate(trackMetadata.getBitrate());
> > -    setReplayGain(trackMetadata.getReplayGain());
> > +void TrackInfoObject::setTrackMetadata(
> > +        const Mixxx::TrackMetadata& trackMetadata,
> > +        QImage *pCoverArt,
> > +        bool parsedFromFile) {
> > +    {
> > +        // enter locking scope
> > +        QMutexLocker lock(&m_qMutex);
> > +
> > +        bool modified = compareAndSet(&m_bHeaderParsed, parsedFromFile);
> > +        bool modifiedReplayGain =
> > +                (m_metadata.getReplayGain() != trackMetadata.getReplayGain());
> > +        bool modifiedCoverArt = false;
> > +        if (m_metadata != trackMetadata) {
> > +            const Mixxx::ReplayGain replayGainBackup(m_metadata.getReplayGain());
>
> Why we need a special treatment for replay gain here? Don't we trust the 
> file replay gain?
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/mixxxdj/mixxx/pull/900/files#r54356482>.
>

uklotzde replied:
Looks like legacy code that survived the introduction of the ReplayGain 
class. Deleted.

On 02/28/2016 10:15 PM, Daniel Schürmann wrote:
>
> In src/trackinfoobject.cpp 
> <https://github.com/mixxxdj/mixxx/pull/900#discussion_r54356482>:
>
> > -    setBitrate(trackMetadata.getBitrate());
> > -    setReplayGain(trackMetadata.getReplayGain());
> > +void TrackInfoObject::setTrackMetadata(
> > +        const Mixxx::TrackMetadata& trackMetadata,
> > +        QImage *pCoverArt,
> > +        bool parsedFromFile) {
> > +    {
> > +        // enter locking scope
> > +        QMutexLocker lock(&m_qMutex);
> > +
> > +        bool modified = compareAndSet(&m_bHeaderParsed, parsedFromFile);
> > +        bool modifiedReplayGain =
> > +                (m_metadata.getReplayGain() != trackMetadata.getReplayGain());
> > +        bool modifiedCoverArt = false;
> > +        if (m_metadata != trackMetadata) {
> > +            const Mixxx::ReplayGain replayGainBackup(m_metadata.getReplayGain());
>
> Why we need a special treatment for replay gain here? Don't we trust the 
> file replay gain?
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/mixxxdj/mixxx/pull/900/files#r54356482>.
>

daschuer commented on src/trackinfoobject.cpp:
```diff
@@ -110,26 +101,49 @@ void TrackInfoObject::setDeleteOnReferenceExpiration(bool deleteOnReferenceExpir
+                m_coverArt.image);
+            m_coverArt.info.coverLocation = QString();
+            m_coverArt.info.type = CoverInfo::METADATA;
+            m_coverArt.info.source = CoverInfo::GUESSED;
```
normally only the caller can set these values.

daschuer replied:
This works here, because setTrackMetadata is only called from SoundSourceProxy::loadTrackMetadataAndCoverArt, and relies on pCoverArt being null when we have already an alternative cover set. 
I think this should moved probably to readTrackMetadataAndCoverArtFromFile()

uklotzde committed:
Decouple setting of track metadata and cover art

uklotzde committed:
Remove special case handling for replay gain

In the past setting the replay gain to an invalid values was
silently ignored. This is no longer necessary since the simple
double value has been replaced by the wrapper class ReplayGain.

uklotzde committed:
Fix declaration of local variable

daschuer commented:
LGTM. Thank you very much.
